internal/configconverter/config_server.go

```diff
@@ -43,117 +46,134 @@ const (
+	mux.HandleFunc(effectivePath, effectiveHandleFunc)
+
+	cs.server = &http.Server{
+		ReadHeaderTimeout: 20 * time.Second,
+		Handler:           mux,
+	}
 	return cs
 }
 
-func (cs *configServer) setInitial(config map[string]any) {
-	cs.initialConfigSync.Lock()
-	cs.initial = config
-	cs.initialConfigSync.Unlock()
+// Convert is intended to be called as the final service confmap.Converter,
+// which registers the service config before being finally resolved and unmarshalled.
+func (cs *ConfigServer) Convert(_ context.Context, conf *confmap.Conf) error {
+	cs.start()
+	cs.setEffective(conf.ToStringMap())
+	return nil
 }
 
-func (cs *configServer) setEffective(config map[string]any) {
-	cs.effectiveConfigSync.Lock()
-	cs.effective = config
-	cs.effectiveConfigSync.Unlock()
+func (cs *ConfigServer) Register() {
+	cs.wg.Add(1)
+}
+
+func (cs *ConfigServer) SetForScheme(scheme string, config map[string]any) {
+	cs.initialMutex.Lock()
+	defer cs.initialMutex.Unlock()
+	cs.initial[scheme] = config
 }
 
-func (cs *configServer) getInitial() map[string]any {
-	cs.initialConfigSync.RLock()
-	defer cs.initialConfigSync.RUnlock()
+func (cs *ConfigServer) getInitial() map[string]any {
+	cs.initialMutex.RLock()
+	defer cs.initialMutex.RUnlock()
 	return cs.initial
 }
 
-func (cs *configServer) getEffective() map[string]any {
-	cs.effectiveConfigSync.RLock()
-	defer cs.effectiveConfigSync.RUnlock()
+func (cs *ConfigServer) setEffective(config map[string]any) {
+	cs.effectiveMutex.Lock()
+	defer cs.effectiveMutex.Unlock()
+	cs.effective = config
+}
+
+func (cs *ConfigServer) getEffective() map[string]any {
+	cs.effectiveMutex.RLock()
+	defer cs.effectiveMutex.RUnlock()
 	return cs.effective
 }
 
-func (cs *configServer) start() error {
+// start will create and start the singleton http server. It presumes cs.Register() has been
+// called at least once and will tear down the moment the final cs.Unregister() call is made.
+func (cs *ConfigServer) start() {
 	if enabled := os.Getenv(configServerEnabledEnvVar); enabled != "true" {
 		// The config server needs to be explicitly enabled for the time being.
-		return nil
+		return
 	}
 
-	endpoint := defaultConfigServerEndpoint
-	if portOverride, ok := os.LookupEnv(configServerPortEnvVar); ok {
-		if portOverride == "" {
-			// If explicitly set to empty do not start the server.
-			return nil
-		}
+	cs.once.Do(
+		func() {
+			endpoint := defaultConfigServerEndpoint
+			if portOverride, ok := os.LookupEnv(configServerPortEnvVar); ok {
+				if portOverride == "" {
+					// If explicitly set to empty do not start the server.
+					return
+				}
 
-		endpoint = "localhost:" + portOverride
-	}
+				endpoint = "localhost:" + portOverride
+			}
 
-	listener, err := net.Listen("tcp", endpoint)
-	if err != nil {
-		return err
-	}
+			listener, err := net.Listen("tcp", endpoint)
+			if err != nil {
+				log.Print(fmt.Errorf("error starting config server: %w", err).Error())
```

Comment:
Should it crash the collector startup instead?

Reply:
I see, so until we have a better way to manage those conflicts, we should instead ignore this error. Understood.

Reply:
Updated to log port-setting guidance on conflicts.